### ISSUE TITLE
hlsl-out miscompilation

### ISSUE THREAD

**Issue by Jasper-Bekkers** [gfx-rs/naga#2471]
Input was roughly

```wglsl
let PIXELS_PER_THREAD = 4u;

fn fill_path(tile: Tile, xy: vec2<f32>, even_odd: bool) -> array<f32, PIXELS_PER_THREAD> {
    var area: array<f32, PIXELS_PER_THREAD>;
...
    return area;
}
```

Generated output was roughly (notice the invalid return type)
```hlsl
float fill_path(Bindings bnd, Tile tile, float2 xy, bool even_odd)
{
    float area_1[4] = (float[4])0;
...
     return area_1;
}
```

Expected (notice: `[4]` at the end there).
```hlsl
float fill_path(Bindings bnd, Tile tile, float2 xy, bool even_odd)[4] {
...
}
```

**Comment by waywardmonkeys** [comment]
I think this is the same as #1930.

Do you have a reference for the syntax that you gave with the `[4]` at the end? That doesn't appear in HLSL docs that I've seen.

I have a patch that might fix this, but need to write tests and understand a couple of things.

**Comment by Jasper-Bekkers** [comment]
> Do you have a reference for the syntax that you gave with the [4] at the end? That doesn't appear in HLSL docs that I've seen.

I've asked around a bit, no idea where I got the syntax form originally but it's supported at least in DXC since forever.

An alternative would be to make it into an out-variable instead.

**Comment by teoxoy** [comment]
This is indeed a duplicate of https://github.com/gfx-rs/naga/issues/1930.

> I've asked around a bit, no idea where I got the syntax form originally but it's supported at least in DXC since forever.

While it works in DXC, FXC doesn't support this syntax.
The only way I found to return an array is to define the type as a `typedef` and use that as the return type.

The fact that the array size always goes after the name of variables/arguments/members is quite annoying.
There are a bunch of places in the backend where we have to account for this.

It might be easier to output `typedef`s for all arrays and use those instead.

**Comment by Jasper-Bekkers** [comment]
@teoxoy the alternative that spirv-cross uses is to turn it into an out variable instead

**Comment by teoxoy** [comment]
I don't think that would be easier to implement. Would it have other benefits compared to the `typedef` approach?
